File: bindings/go/binding_test.go
package tree_sitter_php_test

import (
	"strings"
	"testing"

	tree_sitter "github.com/tree-sitter/go-tree-sitter"
	tree_sitter_php "github.com/tree-sitter/tree-sitter-php/bindings/go"
)

func TestPHPGrammar(t *testing.T) {
	language := tree_sitter.NewLanguage(tree_sitter_php.LanguagePHP())
	if language == nil {
		t.Errorf("Error loading PHP grammar")
	}

	sourceCode := []byte("<?php echo 'Hello, World!';")
	parser := tree_sitter.NewParser()
	defer parser.Close()
	parser.SetLanguage(language)

	tree := parser.Parse(sourceCode, nil)
	if tree == nil || tree.RootNode().HasError() {
		t.Errorf("Error parsing PHP")
	}
}

func TestPHPOnlyGrammar(t *testing.T) {
	language := tree_sitter.NewLanguage(tree_sitter_php.LanguagePHPOnly())
	if language == nil {
		t.Errorf("Error loading PHP-Only grammar")
	}

	sourceCode := []byte("echo 'Hello, World!';")
	parser := tree_sitter.NewParser()
	defer parser.Close()
	parser.SetLanguage(language)

	tree := parser.Parse(sourceCode, nil)
	if tree == nil || tree.RootNode().HasError() {
		t.Errorf("Error parsing PHP")
	}
}

func sequentialHeredocs(prefix string, count int, tag string) []byte {
	block := "$a = <<<" + tag + "\nx\n" + tag + ";\n"
	src := make([]byte, 0, len(prefix)+len(block)*count)
	src = append(src, prefix...)
	for i := 0; i < count; i++ {
		src = append(src, block...)
	}
	return src
}

func parseWithoutError(t *testing.T, language *tree_sitter.Language, source []byte) {
	t.Helper()
	parser := tree_sitter.NewParser()
	defer parser.Close()
	parser.SetLanguage(language)
	tree := parser.Parse(source, nil)
	if tree == nil || tree.RootNode().HasError() {
		t.Errorf("Error parsing sequential heredocs")
	}
}

func TestManySequentialHeredocs(t *testing.T) {
	php := tree_sitter.NewLanguage(tree_sitter_php.LanguagePHP())
	phpOnly := tree_sitter.NewLanguage(tree_sitter_php.LanguagePHPOnly())
	longTag := strings.Repeat("HEREDOC_", 30)
	parseWithoutError(t, php, sequentialHeredocs("<?php\n", 20, longTag))
	parseWithoutError(t, php, sequentialHeredocs("<?php\n", 250, "EOD"))
	parseWithoutError(t, phpOnly, sequentialHeredocs("", 20, longTag))
	parseWithoutError(t, phpOnly, sequentialHeredocs("", 250, "EOD"))
}

File: bindings/node/binding_test.js
/// <reference types="node" />

const assert = require("node:assert");
const { describe, it } = require("node:test");

const Parser = require("tree-sitter");
const { php, php_only } = require("../..");

describe("PHP", () => {
  const parser = new Parser();
  parser.setLanguage(php);

  it("should be named php", () => {
    assert.strictEqual(parser.getLanguage().name, "php");
  });

  it("should parse source code", () => {
    const sourceCode = "<?php echo 'Hello, World!';";
    const tree = parser.parse(sourceCode);
    assert(!tree.rootNode.hasError);
  });

  it("should parse many sequential heredocs", () => {
    const longTag = "HEREDOC_".repeat(30);
    const long = "<?php\n" + `$a = <<<${longTag}\nx\n${longTag};\n`.repeat(20);
    const short = "<?php\n" + "$a = <<<EOD\nx\nEOD;\n".repeat(250);
    assert(!parser.parse(long).rootNode.hasError);
    assert(!parser.parse(short).rootNode.hasError);
  });
});

describe("PHP Only", () => {
  const parser = new Parser();
  parser.setLanguage(php_only);

  it("should be named php_only", () => {
    assert.strictEqual(parser.getLanguage().name, "php_only");
  });

  it("should parse source code", () => {
    const sourceCode = "echo 'Hello, World!';";
    const tree = parser.parse(sourceCode);
    assert(!tree.rootNode.hasError);
  });

  it("should parse many sequential heredocs", () => {
    const longTag = "HEREDOC_".repeat(30);
    const long = `$a = <<<${longTag}\nx\n${longTag};\n`.repeat(20);
    const short = "$a = <<<EOD\nx\nEOD;\n".repeat(250);
    assert(!parser.parse(long).rootNode.hasError);
    assert(!parser.parse(short).rootNode.hasError);
  });
});
